django/db/models/fields/composite.py

```diff
@@ -128,6 +130,25 @@ def _check_field_name(self):
             )
         ]
 
+    def to_json(self, pk, connection):
+        from django.core.serializers.json import DjangoJSONEncoder
+
+        assert len(self.fields) == len(pk)
```

Comment:
Could this ever raise in normal usage? Or is this just a guard against someone misusing an internal API?

I'm wondering if this should be a different exception type that can't be optimised away.

Reply:
It was intended to be more of a guard. The assert in the `from_json` could happen if someone modifies the primary key fields of a modal but doesn't update the related `object_id`.

Reply:
Thinking about this more, I think this should be a full exception, perhaps `ImproperlyConfigured`. I don't think it should be an `assert` unless it's there only for Django contributors.

Reply:
This would only happen at runtime though, when the db is queried. How about FieldError?

Reply:
a ValueError is appropriate here too I think

Reply:
I improved the errors, please check